awscli/customizations/cloudtrail/validation.py

```diff
@@ -555,21 +551,29 @@ def validate(self, bucket, key, public_key, digest_data, inflated_digest):
         """
         try:
             decoded_key = base64.b64decode(public_key)
-            public_key = load_der_public_key(decoded_key, default_backend())
-            to_sign = self._create_string_to_sign(digest_data, inflated_digest)
-            signature_bytes = binascii.unhexlify(digest_data['_signature'])
-            hashing = hashes.SHA256()
-            public_key.verify(signature_bytes, to_sign, PKCS1v15(), hashing)
-        except ValueError:
+            public_key = RSA.new_public_key_from_der_data(decoded_key)
+        except RuntimeError:
             raise DigestError(
                 ('Digest file\ts3://%s/%s\tINVALID: Unable to load PKCS #1 key'
                  ' with fingerprint %s')
                 % (bucket, key, digest_data['digestPublicKeyFingerprint']))
-        except cryptography.exceptions.InvalidSignature:
-            # Don't display the stack trace of a signature exception to avoid
+
+        to_sign = self._create_string_to_sign(digest_data, inflated_digest)
+        signature_bytes = binascii.unhexlify(digest_data['_signature'])
+
+
+        result = public_key.verify(
+            signature_algorithm=RSASignatureAlgorithm.PKCS1_5_SHA256, 
+            digest=hashlib.sha256(to_sign).digest(), 
+            signature=signature_bytes
+        )
+        if not result:
+            # The previous implementation caught a cryptography.exceptions.InvalidSignature
+            # exception here and re-raised the DigestSignatureError to avoid the stack trace
             # leaking any information about the key.
```

Comment:
Does CRT throw any errors we'd want the same behavior for?

Reply:
All errors are raised as a `RuntimeError`. Examples:

### malformed key file loading

```
Traceback (most recent call last):
  File "/Users/kdaily/issues/use-crt-cryptography/test_pem_private_key.py", line 8, in <module>
    RSA.new_private_key_from_pem_data(key_bytes)
  File "/Users/kdaily/issues/use-crt-cryptography/venv/lib/python3.12/site-packages/awscrt/crypto.py", line 116, in new_private_key_from_pem_data
    return RSA(binding=_awscrt.rsa_private_key_from_pem_data(pem_data))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: 1179 (AWS_ERROR_PEM_MALFORMED): Malformed PEM object encountered.
```

You get typing errors if you pass the wrong type (e.g., strings instead of bytes to the `signature` parameter):

```
Traceback (most recent call last):
  File "/Users/kdaily/issues/use-crt-cryptography/test_pem_private_key.py", line 25, in <module>
    public_key.verify(
  File "/Users/kdaily/issues/use-crt-cryptography/venv/lib/python3.12/site-packages/awscrt/crypto.py", line 173, in verify
    return _awscrt.rsa_verify(self._binding, signature_algorithm, digest, signature)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: a bytes-like object is required, not 'str'
```

See a comment above similarly about errors for loading the public key file.

Reply:
When verifying, I tried changing the hashing/padding algorithms, and I just get back `False` (not verified), no errors.

Reply:
Similarly when signing, I tried changing the hash used and it still returned a signature with no error, like:

```
signature = private_key.sign(
    RSASignatureAlgorithm.PKCS1_5_SHA1,                # Declare using SHA1
    digest=hashlib.sha256(message.encode()).digest()   # But actually hash with SHA256
)
```